ci(release): GHA release workflow

megglos opened the pull request:
## Description

This introduces a new GHA release workflow, replicating all steps of the previous jenkins release job as defined in #11142.

Functional changes in comparison to previous Jenkins job:
- a new optional input parameter `releaseBranch` was added which allows to release from an arbitrary branch, I decided to add this to make testing easier
- the `PUSH_DOCKER` input parameter was omitted, instead all write behavior to "remote" is controlled with a single input being `dryRun` that defaults to true

I split the release into three jobs:
- first being the intertwined Maven & Go release (might be worth a follow-up to decouple if feasible) producing the release artifacts and attaching them to the workflow run
- second, depending on the release job, being the creation of the Github release, also generating the checksum files
- third, depending on the release job, being the docker image build, verification & push

I would like to perform some more "hot" releases (pushChanges=true) before we merge, but would like to get review feedback first.

Dry Test runs:

- [alpha release `0.0.4-alpha5`](https://github.com/camunda/zeebe/actions/runs/3677179115)
- [stable release `0.0.4` without tagging of latest image](https://github.com/camunda/zeebe/actions/runs/3677286634)
- [stable release `0.0.4` with tagging of latest image](https://github.com/camunda/zeebe/actions/runs/3677185231)
- [failed release job, sending a slack notification](https://github.com/camunda/zeebe/actions/runs/3677141190) and the [corresponding slack message](https://camunda.slack.com/archives/C013MEVQ4M9/p1670858299585059)

**Hot test runs** before merge (using a `meg-release-test` branch based on this branch):

- [x] release an alpha version `0.0.4-alpha1`: [job](https://github.com/camunda/zeebe/actions/runs/3683289329/jobs/6231911360), [release notes](https://github.com/camunda/zeebe/releases/tag/untagged-605833993b331e117379), [tag](https://github.com/camunda/zeebe/releases/tag/0.0.4-alpha1) , [docker image](https://hub.docker.com/layers/camunda/zeebe/0.0.4-alpha1/images/sha256-dec341e8ee187c994fa186b458e43d5383f09fbbc3df2dd075762a64885be189?context=explore)
- [x] release a stable version `0.0.4`: [job](https://github.com/camunda/zeebe/actions/runs/3683416289), [release notes](https://github.com/camunda/zeebe/releases/tag/untagged-108ebe3275f619f8d3cc), [tag](https://github.com/camunda/zeebe/releases/tag/0.0.4), [docker image](https://hub.docker.com/layers/camunda/zeebe/0.0.4/images/sha256-60ed0640f53b65c241a75dc22d08309b4b7f04d12333d418b9e7f74327c5795e?context=explore)

## Related issues

<!-- Which issues are closed by this PR or are related -->

closes #11142

<!-- Cut-off marker
_All lines under and including the cut-off marker will be removed from the merge commit message_

## Definition of Ready

Please check the items that apply, before requesting a review.

You can find more details about these items in our wiki page about [Pull Requests and Code Reviews](https://github.com/camunda/zeebe/wiki/Pull-Requests-and-Code-Reviews).

* [ ] I've reviewed my own code
* [ ] I've written a clear changelist description
* [ ] I've narrowly scoped my changes
* [ ] I've separated structural from behavioural changes
-->

## Definition of Done

<!-- Please check the items that apply, before merging or (if possible) before requesting a review. -->

Testing:
* [ ] There are unit/integration tests that verify all acceptance criterias of the issue
* [ ] New tests are written to ensure backwards compatibility with further versions
* [x] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark

Please refer to our [review guidelines](https://github.com/camunda/zeebe/wiki/Pull-Requests-and-Code-Reviews#code-review-guidelines).

oleschoenburg commented on .github/workflows/release.yml:
```diff
@@ -0,0 +1,340 @@
+name: Release
```
:x: Please add a concurrency config like we do for [deploying snapshots](https://github.com/camunda/zeebe/blob/38d8edf95746f4060d725ad5771ed89f17db2a34/.github/workflows/ci.yml#L494-L496). This ensures that we don't produce a mess when two releases are triggered concurrently.

megglos replied:
wdyt about doing this based on the release-branch used? that would still allow e.g. doing multiple patch releases in parallel which seems reasonable to do?

oleschoenburg replied:
Yeah that seems reasonable :+1:

megglos replied:
https://github.com/camunda/zeebe/pull/11240/commits/1d33d49699eb747aa67b56910117cdb6ddee532c

oleschoenburg commented on .github/actions/build-docker/action.yml:
```diff
@@ -74,14 +85,14 @@ runs:
-          REVISION=${{ github.sha }}
-          VERSION=${{ steps.get-version.outputs.result }}
+          REVISION=${{ inputs.revision != '' && inputs.revision || github.sha }}
+          VERSION=${{ inputs.version != '' && inputs.version || steps.get-version.outputs.result }}
```
:x: Using `steps.get-version.outputs.result` as a fallback seems wrong as that might be something like `camunda/zeebe:8.2.0, camunda/zeebe:latest`

megglos replied:
seems you may have confused that with the `get-images` step? the output of `get-version` is just the plain version [either from input or determined via mvn command](https://github.com/camunda/zeebe/pull/11240/files/cd04e42c60937a2051023556f2beafced28dfb79#diff-9288d890756492a0e0289f15c477f58821f3a2d9af4274386b09003a4ef57354R67)

oleschoenburg replied:
Oops, yeah you are right :+1:

oleschoenburg commented on .github/actions/build-docker/action.yml:
```diff
@@ -56,16 +62,21 @@ runs:
       id: get-version
+      if: ${{ inputs.version == ''}}
       shell: bash
       run: echo ::set-output name=result::$(mvn -q -Dexec.executable=echo -Dexec.args='${project.version}' --non-recursive exec:exec)
```
:wrench: use `GITHUB_OUTPUT` here

megglos commented:
> Do you need my review as well, or is this just for sharing knowledge?

totally fine if you only manage to look later into it for knowledge sharing + would appreciate any suggestions that may come to your mind then.
Given Ole approves after I addressed the comments I would already move forward then.

megglos commented on .github/workflows/release.yml:
```diff
@@ -0,0 +1,340 @@
+      isLatest:
+        description: 'Whether this is the latest release and the docker image should be tagged as camunda/zeebe:latest'
+        required: false
+      pushChanges:
```
https://github.com/camunda/zeebe/pull/11240/commits/620eadc84ed217df4b163e7d82c8854df198e2e8

oleschoenburg reviewed (approved):
:rocket:

megglos commented on .github/workflows/release.yml:
```diff
@@ -0,0 +1,349 @@
+        description: 'Whether to push release commits and artifacts to remote (Git commits & tags, Maven artifacts, Github Release), defaults to true.'
+        default: true
+
+concurrency:
```
https://github.com/camunda/zeebe/pull/11240/commits/e6392024417cbd7d07d5db4752c217219cbb147a

npepinpe commented on .github/workflows/release.yml:
```diff
@@ -0,0 +1,349 @@
+on:
+  workflow_dispatch:
+    inputs:
+      releaseBranch:
```
:question: How is this treated without type information? e.g. option, boolean, etc. Are they all strings? How will it look like in the dispatch form? :thinking:

megglos replied:
default is string according to https://github.blog/changelog/2021-11-10-github-actions-input-types-for-manual-workflows/

official docs don't specify this well 🤷 
https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#workflow_dispatch

npepinpe commented on .github/workflows/release.yml:
```diff
@@ -0,0 +1,349 @@
+jobs:
+  release:
+    name: Maven & Go Release
+    runs-on: n1-standard-16-netssd-preempt-quick
```
:question: Could we have non pre-emptible nodes for the release? It would suck to wait 20 minutes for it to fail because it gets trashed :smile:

megglos replied:
will create a follow-up to setup such a runner pool

megglos replied:
added [here](https://github.com/camunda/zeebe/issues/10992)

megglos commented:
Successfully performed two hot releases alpha and stable 🎉 , see checked test cases in the description. Will merge.

megglos commented:
bors r+
